hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -324,6 +319,8 @@ def cached_test_function(self, buffer):
             pass
         result = self.__engine.cached_test_function(buffer)
         self.incorporate_test_data(result)
+        if result.status == Status.OVERRUN:
+            result = OverrunPlaceholder(result.buffer)
```

Comment:
If this replacement takes place after the initial `incorporate`, then I think we can end up with an unpredictable mixture of real and fake overruns in the shrinker's cache.

It might be best to try to enforce the invariant that the rest of the shrinker never gets to see a real overrun under any circumstances. That will miss some corner-case shrinks, but at least it will do so consistently, instead of being extremely sensitive to the precise history of the cache.

(If so, then we should add a corresponding assertion to `incorporate`, so that it never accidentally stores a real overrun.)

Reply:
A related option is to move the placeholder-replacement into the engine itself.

That way no part of the shrinker ever sees a real overrun.

Reply:
Good call. I've modified `cached_test_function` in the runner to always return `Overrun` in preference to a real overrun.